header.go

```diff
@@ -101,19 +107,17 @@ func readHeader(r *bufio.Reader, p *Part) (textproto.MIMEHeader, error) {
 		firstColon := bytes.IndexByte(s, ':')
 		firstSpace := bytes.IndexAny(s, " \t\n\r")
 		if firstSpace == 0 {
-			// If there is a single colon followed immediately by a space and there
-			// is no equal sign, or comes before any equal sign, then it should not
-			// be considered as a continuation.
-			colonSpaceIdx := bytes.Index(s, []byte{':', ' '})
-			equalIdx := bytes.IndexByte(s, '=')
-			if bytes.Count(s, []byte{':'}) == 1 && colonSpaceIdx != -1 &&
-				(equalIdx == -1 || equalIdx > colonSpaceIdx) {
-				s = textproto.TrimBytes(s)
+			// If the line begins with some space, followed by a header-like
+			// string (any combination of upper and lower case letters,
+			// numbers and dash sign), then it should not be considered
+			// as a continuation but as a new header.
+			sTrimmed := textproto.TrimBytes(s)
+			if firstSpace < firstColon && headerDeclarationRegex.Match(sTrimmed) {
```

Comment:
This is actually quite a bit more restrictive than the logic that was there before, which didn't restrict header names to `a-zA-Z0-9-`. I think this should continue to be as lenient.

Something more like:
```suggestion
			// Get the index of the first whitespace (if any) after the apparent folding whitespace
			firstSpace := bytes.IndexAny(sTrimmed, " \t\n\r")
			if firstSpace > 0 && firstSpace < firstColon {
```

I haven't tested that, but that's the jist of the idea.

Reply:
I agree with Daniel, would prefer to avoid the regex.

Reply:
I partially agree, I am ok in removing the regex and doing some 'magic' in relative positions between space and colon, but would still keep the constraint on the first space right after the colon. This is to avoid problems with i.e. DKIM signatures, but I will update the PR so that it's clear what I mean.

Reply:
> keep the constraint on the first space right after the colon.

I'm not sure that's a good idea. I expect headers without a space after the `:` are often encountered.

I also checked [the RFC](https://tools.ietf.org/html/rfc5322) to see what it said about whether that space is even required, and couldn't find a mention of it (someone please prove me wrong).

But to make it even worse, I found [this gem](https://tools.ietf.org/html/rfc5322#section-4.5) (regarding headers):

> Also, any amount of white space is allowed before the ":" at the end of the field name.

So I guess the logic needs to allow non-folding white space as long as it's right before the ':' of the header.

Ugh.

Reply:
Don't get me wrong, this is not per RFC of course, no space is required in a header declaration. 
But as well what we are trying to achieve here is not RFC compliant, so this is being another best-effort attempt to parse some headers as new declarations whereas, according to RFC, should be considered as folded headers.

> Also, any amount of white space is allowed before the ":" at the end of the field name.
This is a good one, and of course would fail with the code here. Nice.

To recap a bit, since this is already an out of standard parsing attempt, I think the less damage solution should be implemented, that means (my personal opinion about this):
 since these headers should not be treated as new at all, if the library manages to cover the most commons header declarations without breaking the parsing of standard headers, it would be a win.

I think the code which is currently on master is worse than the proposed solutions, because it's looking for an '=' sign, attempting to fix only some very specific cases. But again, this is just my personal point of view :)

Reply:
I agree the equals sign detection is probably not what we want.  I don't have much time to think or review until the weekend.

Reply:
Any feedback about this? Thanks! :)